bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/discovery/CaddxDiscoveryService.java

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.caddx.internal.discovery;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.caddx.internal.CaddxBindingConstants;
+import org.openhab.binding.caddx.internal.CaddxEvent;
+import org.openhab.binding.caddx.internal.config.CaddxKeypadConfiguration;
+import org.openhab.binding.caddx.internal.config.CaddxPartitionConfiguration;
+import org.openhab.binding.caddx.internal.config.CaddxZoneConfiguration;
+import org.openhab.binding.caddx.internal.handler.CaddxBridgeHandler;
+import org.openhab.binding.caddx.internal.handler.CaddxThingType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for discovering the supported Things.
+ *
+ * @author Georgios Moutsos - Initial contribution
+ */
+@NonNullByDefault
+public class CaddxDiscoveryService extends AbstractDiscoveryService {
```

Comment:
The BaseThingHandlerFactory has a feature that will automatically create and register ThingHandlerService classes as osgi services when their associated handler is created. All you would need to do to take advantage of this is to implement ThingHandlerService and the list of interfaces you wish it to register as.
```suggestion
public class CaddxDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService, DiscoveryService {
```

Then you will need to replace your CaddxDiscoveryService constructor with a no-arg constructor, otherwise the BaseThingHandlerFactory can't create your CaddxDiscoveryService.

And lastly in your handler class you just override the `getServices` method. 
```
public CaddxBridgeHandler {
...
public Collection<Class<? extends ThingHandlerService>> getServices() {
        return Collections.singleton(CaddxDiscoveryService.class);
    }
...
}
```

Then you are off to the races.
Doing this should let you get rid of all the `discoveryServiceRegistrations` related code in your handler factory since it will be handled for you automatically.

Reply:
I will look into it

Reply:
Done